File: game-server/src/main/java/work/fking/pangya/game/SessionClient.java
package work.fking.pangya.game;

import com.fasterxml.jackson.databind.ObjectMapper;
import io.lettuce.core.RedisClient;
import io.lettuce.core.api.sync.RedisCommands;

import java.io.IOException;

public class SessionClient {

    private static final ObjectMapper OBJECT_MAPPER = new ObjectMapper();
    private final RedisCommands<String, String> redisCommands;

    private SessionClient(RedisCommands<String, String> redisCommands) {
        this.redisCommands = redisCommands;
    }

    public static SessionClient create(RedisClient redisClient) {
        return new SessionClient(redisClient.connect().sync());
    }

    public SessionInfo loadSession(String sessionKey) throws IOException {
        var json = redisCommands.get(sessionKey);
        if (json == null) {
            return null;
        }
        return OBJECT_MAPPER.readValue(json, SessionInfo.class);
    }

    public record SessionInfo(
            String sessionKey,
            int uid,
            String username,
            String nickname
    ) {

    }
}

File: game-server/src/main/kotlin/work/fking/pangya/game/SessionClient.kt
package work.fking.pangya.game

import com.fasterxml.jackson.module.kotlin.jacksonObjectMapper
import com.fasterxml.jackson.module.kotlin.readValue
import io.lettuce.core.RedisClient
import io.lettuce.core.api.sync.RedisCommands
import work.fking.pangya.game.player.Player

class SessionClient(
    private val redis: RedisCommands<String, String>
) {
    constructor(redisClient: RedisClient) : this(redisClient.connect().sync())

    private val objectMapper = jacksonObjectMapper()

    fun unregisterSession(player: Player) {
        redis.del("session-${player.username}")
    }

    fun loadHandoverInfo(loginKey: String): HandoverInfo? {
        val json = redis[loginKey] ?: return null
        return objectMapper.readValue<HandoverInfo>(json)
    }

    @JvmRecord
    data class HandoverInfo(
        val targetServerId: Int,
        val loginKey: String,
        val uid: Int,
        val username: String,
        val nickname: String,
        val characterIffId: Int?,
        val characterHairColor: Int?
    )
}
